common/persistence/sql/queue_v2.go

```diff
@@ -26,36 +26,183 @@ package sql
 
 import (
 	"context"
-	"errors"
+	"database/sql"
 	"fmt"
 
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/api/serviceerror"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
 	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/sql/sqlplugin"
 )
 
-type (
-	queueV2 struct{}
+const (
+	EmptyPartition = 0
+	// pageTokenPrefixByte is the first byte of the serialized page token. It's used to ensure that the page token is
+	// not empty. Without this, if the last_read_message_id is 0, the serialized page token would be empty, and clients
+	// could erroneously assume that there are no more messages beyond the first page. This is purely used to ensure
+	// that tokens are non-empty; it is not used to verify that the token is valid like the magic byte in some other
+	// protocols.
+	pageTokenPrefixByte = 0
 )
 
-var (
-	ErrNotImplemented = errors.New("method is not implemented yet")
+type (
+	queueV2 struct {
+		logger log.Logger
+		SqlStore
+	}
+
+	QueueV2MetadataPayload struct {
+		AckLevel int64
+	}
 )
 
-// NewQueueV2 returns an implementation of [persistence.QueueV2] which always returns an error because it is not
-// implemented yet.
-func NewQueueV2() persistence.QueueV2 {
-	return &queueV2{}
+// NewQueueV2 returns an implementation of persistence.QueueV2.
+func NewQueueV2(db sqlplugin.DB,
+	logger log.Logger) (persistence.QueueV2, error) {
+
+	queue := &queueV2{
+		SqlStore: NewSqlStore(db, logger),
+		logger:   logger,
+	}
+	return queue, nil
 }
 
 func (q queueV2) EnqueueMessage(
-	context.Context,
-	*persistence.InternalEnqueueMessageRequest,
+	ctx context.Context,
+	request *persistence.InternalEnqueueMessageRequest,
 ) (*persistence.InternalEnqueueMessageResponse, error) {
-	return nil, fmt.Errorf("%w: EnqueueMessage", ErrNotImplemented)
+	var lastMessageID int64
+	tx, err := q.Db.BeginTx(ctx)
+	if err != nil {
+		return nil, serviceerror.NewUnavailable(fmt.Sprintf("EnqueueMessage failed. Failed to start transaction. Error: %v", err))
+	}
+	lastMessageID, err = tx.GetLastEnqueuedMessageIDForUpdateV2(ctx, sqlplugin.QueueV2Filter{
+		QueueType: request.QueueType,
+		QueueName: request.QueueName,
+		Partition: EmptyPartition,
+	})
+	switch err {
+	case nil:
+		lastMessageID = lastMessageID + 1
+	case sql.ErrNoRows:
+		lastMessageID = persistence.FirstQueueMessageID
+	default:
+		rollBackErr := tx.Rollback()
+		if rollBackErr != nil {
+			q.logger.Error("transaction rollback error", tag.Error(rollBackErr))
+		}
+		return nil, fmt.Errorf("failed to get last enqueued message id: %w", err)
+	}
+	_, err = tx.InsertIntoQueueV2Messages(ctx, []sqlplugin.QueueV2MessageRow{
+		newQueueV2Row(request.QueueType, request.QueueName, lastMessageID, request.Blob),
+	})
+	if err != nil {
+		rollBackErr := tx.Rollback()
+		if rollBackErr != nil {
+			q.logger.Error("transaction rollback error", tag.Error(rollBackErr))
+		}
```

Comment:
I think we should add a return here. Also, would it be possible for you to add a test for this case? I think we can have some SQL-specific test case in addition to the generic queue ones. You don't need to go 100% coverage like I did with Cassandra, but I think something like this would be good since the control-flow here is a bit unorthodox. You can achieve it by wrapping the DB object with something that produces the specific errors you need to get to this branch.

Create a DB wrapper object and a test suite that uses it:

```go
package sqltest

import (
	"context"
	"database/sql"
	"errors"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	commonpb "go.temporal.io/api/common/v1"
	enumspb "go.temporal.io/api/enums/v1"

	"go.temporal.io/server/common/log"
	"go.temporal.io/server/common/log/tag"
	"go.temporal.io/server/common/persistence"
	persistencesql "go.temporal.io/server/common/persistence/sql"
	"go.temporal.io/server/common/persistence/sql/sqlplugin"
)

type (
	faultyDB struct {
		sqlplugin.DB
		insertErr   error
		rollbackErr error
		commitCalls int
	}
	faultyTx struct {
		db *faultyDB
		sqlplugin.Tx
		commitCalls *int
	}
	logRecorder struct {
		log.Logger
		errMsgs []string
	}
)

func (db *faultyDB) BeginTx(ctx context.Context) (sqlplugin.Tx, error) {
	tx, err := db.DB.BeginTx(ctx)
	if err != nil {
		return nil, err
	}
	return &faultyTx{db: db, commitCalls: &db.commitCalls, Tx: tx}, nil
}

func (tx *faultyTx) InsertIntoQueueV2Messages(ctx context.Context, row []sqlplugin.QueueV2MessageRow) (sql.Result, error) {
	if _, err := tx.Tx.InsertIntoQueueV2Messages(ctx, row); err != nil {
		return nil, err
	}
	return nil, tx.db.insertErr
}

func (tx *faultyTx) Rollback() error {
	if err := tx.Tx.Rollback(); err != nil {
		return err
	}
	return tx.db.rollbackErr
}

func (tx *faultyTx) Commit() error {
	*tx.commitCalls++
	return tx.Tx.Commit()
}

func (l *logRecorder) Error(msg string, _ ...tag.Tag) {
	l.errMsgs = append(l.errMsgs, msg)
}

func TestQueueV2(t *testing.T, baseDB sqlplugin.DB) {
	db := &faultyDB{
		DB:          baseDB,
		insertErr:   errors.New("example insert error"),
		rollbackErr: errors.New("example rollback error"),
	}
	logger := &logRecorder{Logger: log.NewTestLogger()}
	q, err := persistencesql.NewQueueV2(db, logger)
	require.NoError(t, err)
	_, err = q.EnqueueMessage(context.Background(), &persistence.InternalEnqueueMessageRequest{
		QueueType: persistence.QueueTypeHistoryNormal,
		QueueName: "test-queue-" + t.Name(),
		Blob: commonpb.DataBlob{
			EncodingType: enumspb.ENCODING_TYPE_JSON,
			Data:         []byte("1"),
		},
	})
	require.Error(t, err)
	assert.ErrorContains(t, err, "example insert error")
	require.Len(t, logger.errMsgs, 1)
	assert.Contains(t, logger.errMsgs[0], "rollback error")
	assert.Zero(t, db.commitCalls)
}
```

And then invoke it here:
```go
func TestPostgreSQLQueueV2(t *testing.T) {
	testData, tearDown := setUpPostgreSQLTest(t)
	t.Cleanup(tearDown)

	queue, err := testData.Factory.NewQueueV2()
	require.NoError(t, err)
	if err != nil {
		t.Fatalf("unable to create Queue: %v", err)
	}
	t.Run("Generic", func(t *testing.T) {
		t.Parallel()
		RunQueueV2TestSuite(t, queue)
	})
	t.Run("SQL", func(t *testing.T) {
		t.Parallel()

		db, err := testData.Factory.GetDB() // have to add this method
		require.NoError(t, err)
		sqltest.TestQueueV2(t, db)
	})
}
```

Reply:
Thanks for this code snippet. I have added 100% test coverage for SQL implementation. This was super helpful.